File: TrajectoryCalculator.go
package go_ballisticcalc

import (
	"math"

	"github.com/gehtsoft-usa/go_ballisticcalc/bmath/unit"
	"github.com/gehtsoft-usa/go_ballisticcalc/bmath/vector"
)

const cZeroFindingAccuracy float64 = 0.000005
const cMinimumVelocity float64 = 50.0
const cMaximumDrop float64 = -15000
const cMaxIterations int = 10
const cGravityConstant float64 = -32.17405

//TrajectoryCalculator table is used to calculate the trajectory of a projectile shot with the parameters specified
type TrajectoryCalculator struct {
	maximumCalculatorStepSize unit.Distance
}

//MaximumCalculatorStepSize returns the maximum size of one calculation iteration.
func (v TrajectoryCalculator) MaximumCalculatorStepSize() unit.Distance {
	return v.maximumCalculatorStepSize
}

//SetMaximumCalculatorStepSize sets the maximum size of one calculation iteration.
//
//As the generic rule, the maximum step of the calculation must not be greater than
//a half of the step used in the short parameter. The smaller value is, the calculation is more precise but
//takes more time to calculate. From practical standpoint the value in range from 0.5 to 5 feet produces
//good enough accuracy.
func (v *TrajectoryCalculator) SetMaximumCalculatorStepSize(x unit.Distance) {
	v.maximumCalculatorStepSize = x
}

func (v TrajectoryCalculator) getCalculationStep(step float64) float64 {
	step = step / 2 //do it twice for increased accuracy of velocity calculation and 10 times per step

	var maximumStep float64 = v.maximumCalculatorStepSize.In(unit.DistanceFoot)
	if step > maximumStep {

		var stepOrder = int(math.Floor(math.Log10(step)))
		var maximumOrder = int(math.Floor(math.Log10(maximumStep)))

		step = step / math.Pow(10, float64(stepOrder-maximumOrder+1))
	}
	return step
}

//CreateTrajectoryCalculator creates and instance of the trajectory calculator
func CreateTrajectoryCalculator() TrajectoryCalculator {
	return TrajectoryCalculator{
		maximumCalculatorStepSize: unit.MustCreateDistance(1, unit.DistanceFoot),
	}
}

//SightAngle calculates the sight angle for a rifle with scope height specified and zeroed using the ammo specified at
//the range specified and under the conditions (atmosphere) specified.
//
//The calculated value is to be used as sightAngle parameter of the ShotParameters structure
func (v TrajectoryCalculator) SightAngle(ammunition Ammunition, weapon Weapon, atmosphere Atmosphere) unit.Angular {
	var calculationStep = v.getCalculationStep(unit.MustCreateDistance(10, weapon.Zero().ZeroDistance().Units()).In(unit.DistanceFoot))

	var deltaRangeVector, rangeVector, velocityVector, gravityVector vector.Vector
	var muzzleVelocity, velocity, barrelAzimuth, barrelElevation float64
	var densityFactor, mach, drag, zeroFindingError float64
	var time, deltaTime float64
	var maximumRange float64

	mach = atmosphere.Mach().In(unit.VelocityFPS)
	densityFactor = atmosphere.getDensityFactor()
	muzzleVelocity = ammunition.MuzzleVelocity().In(unit.VelocityFPS)
	barrelAzimuth = 0.0
	barrelElevation = 0

	zeroFindingError = cZeroFindingAccuracy * 2
	var iterationsCount int

	gravityVector = vector.Create(0, cGravityConstant, 0)
	for zeroFindingError > cZeroFindingAccuracy && iterationsCount < cMaxIterations {
		velocity = muzzleVelocity
		time = 0.0

		//x - distance towards target,
		//y - drop and
		//z - windage
		rangeVector = vector.Create(0.0, -weapon.SightHeight().In(unit.DistanceFoot), 0)
		velocityVector = vector.Create(math.Cos(barrelElevation)*math.Cos(barrelAzimuth), math.Sin(barrelElevation), math.Cos(barrelElevation)*math.Sin(barrelAzimuth)).MultiplyByConst(velocity)
		var zeroDistance float64 = weapon.Zero().ZeroDistance().In(unit.DistanceFoot)
		maximumRange = zeroDistance + calculationStep

		for rangeVector.X <= maximumRange {
			if velocity < cMinimumVelocity || rangeVector.Y < cMaximumDrop {
				break
			}

			deltaTime = calculationStep / velocityVector.X
			velocity = velocityVector.Magnitude()
			drag = densityFactor * velocity * ammunition.Bullet().BallisticCoefficient().Drag(velocity/mach)
			velocityVector = velocityVector.Subtract((velocityVector.MultiplyByConst(drag).Subtract(gravityVector)).MultiplyByConst(deltaTime))
			deltaRangeVector = vector.Create(calculationStep, velocityVector.Y*deltaTime, velocityVector.Z*deltaTime)
			rangeVector = rangeVector.Add(deltaRangeVector)
			velocity = velocityVector.Magnitude()
			time = time + deltaRangeVector.Magnitude()/velocity

			if math.Abs(rangeVector.X-zeroDistance) < 0.5*calculationStep {
				zeroFindingError = math.Abs(rangeVector.Y)
				barrelElevation = barrelElevation - rangeVector.Y/rangeVector.X
				break
			}
		}
		iterationsCount++
	}
	return unit.MustCreateAngular(barrelElevation, unit.AngularRadian)
}

//Trajectory calculates the trajectory with the parameters specified
func (v TrajectoryCalculator) Trajectory(ammunition Ammunition, weapon Weapon, atmosphere Atmosphere, shotInfo ShotParameters, windInfo []WindInfo) []TrajectoryData {
	var rangeTo float64 = shotInfo.MaximumDistance().In(unit.DistanceFoot)
	var step float64 = shotInfo.Step().In(unit.DistanceFoot)

	var calculationStep = v.getCalculationStep(step)

	var deltaRangeVector, rangeVector, velocityAdjusted, velocityVector, windVector, gravityVector vector.Vector
	var muzzleVelocity, velocity, barrelAzimuth, barrelElevation float64
	var densityFactor, mach, drag float64
	var time, deltaTime float64
	var maximumRange, nextRangeDistance float64
	var bulletWeight float64

	bulletWeight = ammunition.Bullet().BulletWeight().In(unit.WeightGrain)

	var stabilityCoefficient = 1.0
	var calculateDrift bool

	if weapon.HasTwist() && ammunition.Bullet().HasDimensions() {
		stabilityCoefficient = calculateStabilityCoefficient(ammunition, weapon, atmosphere)
		calculateDrift = true
	}

	var rangesLength = int(math.Floor(rangeTo/step)) + 1
	var ranges = make([]TrajectoryData, rangesLength)

	barrelAzimuth = 0.0
	barrelElevation = shotInfo.SightAngle().In(unit.AngularRadian)
	barrelElevation = barrelElevation + shotInfo.ShotAngle().In(unit.AngularRadian)
	var alt0 float64 = atmosphere.Altitude().In(unit.DistanceFoot)
	densityFactor, mach = atmosphere.getDensityFactorAndMachForAltitude(alt0)
	var currentWind int
	var nextWindRange = 1e7

	if len(windInfo) < 1 {
		windVector = vector.Create(0, 0, 0)
	} else {
		if len(windInfo) > 1 {
			nextWindRange = windInfo[0].untilDistance.In(unit.DistanceFoot)
		}
		windVector = windToVector(shotInfo, windInfo[0])
	}

	muzzleVelocity = ammunition.MuzzleVelocity().In(unit.VelocityFPS)
	gravityVector = vector.Create(0, cGravityConstant, 0)
	velocity = muzzleVelocity
	time = 0.0

	//x - distance towards target,
	//y - drop and
	//z - windage
	rangeVector = vector.Create(0.0, -weapon.SightHeight().In(unit.DistanceFoot), 0)
	velocityVector = vector.Create(math.Cos(barrelElevation)*math.Cos(barrelAzimuth), math.Sin(barrelElevation), math.Cos(barrelElevation)*math.Sin(barrelAzimuth)).MultiplyByConst(velocity)

	var currentItem int
	maximumRange = rangeTo
	nextRangeDistance = 0

	var twistCoefficient float64

	if calculateDrift {
		if weapon.Twist().Direction() == TwistLeft {
			twistCoefficient = 1
		} else {
			twistCoefficient = -1
		}
	}

	//run all the way down the range
	for rangeVector.X <= maximumRange+calculationStep {
		if velocity < cMinimumVelocity || rangeVector.Y < cMaximumDrop {
			break
		}

		densityFactor, mach = atmosphere.getDensityFactorAndMachForAltitude(alt0 + rangeVector.Y)
		//densityFactor = atmosphere.DensityFactor()
		//mach = atmosphere.Mach().In(unit.Velocity_FPS)

		if rangeVector.X >= nextWindRange {
			currentWind++
			windVector = windToVector(shotInfo, windInfo[currentWind])

			if currentWind == len(windInfo)-1 {
				nextWindRange = 1e7
			} else {
				nextWindRange = windInfo[currentWind].untilDistance.In(unit.DistanceFoot)
			}
		}

		if rangeVector.X >= nextRangeDistance {
			var windage float64 = rangeVector.Z
			if calculateDrift {
				windage += (1.25 * (stabilityCoefficient + 1.2) * math.Pow(time, 1.83) * twistCoefficient) / 12.0
			}

			var dropAdjustment = getCorrection(rangeVector.X, rangeVector.Y)
			var windageAdjustment = getCorrection(rangeVector.X, windage)

			ranges[currentItem] = TrajectoryData{
				time:              Timespan{time: time},
				travelDistance:    unit.MustCreateDistance(rangeVector.X, unit.DistanceFoot),
				drop:              unit.MustCreateDistance(rangeVector.Y, unit.DistanceFoot),
				dropAdjustment:    unit.MustCreateAngular(dropAdjustment, unit.AngularRadian),
				windage:           unit.MustCreateDistance(windage, unit.DistanceFoot),
				windageAdjustment: unit.MustCreateAngular(windageAdjustment, unit.AngularRadian),
				velocity:          unit.MustCreateVelocity(velocity, unit.VelocityFPS),
				mach:              velocity / mach,
				energy:            unit.MustCreateEnergy(calculateEnergy(bulletWeight, velocity), unit.EnergyFootPound),
				optimalGameWeight: unit.MustCreateWeight(calculateOgv(bulletWeight, velocity), unit.WeightPound),
			}
			nextRangeDistance += step
			currentItem++
			if currentItem == len(ranges) {
				break
			}
		}

		deltaTime = calculationStep / velocityVector.X
		velocityAdjusted = velocityVector.Subtract(windVector)
		velocity = velocityAdjusted.Magnitude()
		drag = densityFactor * velocity * ammunition.Bullet().BallisticCoefficient().Drag(velocity/mach)
		velocityVector = velocityVector.Subtract((velocityAdjusted.MultiplyByConst(drag).Subtract(gravityVector)).MultiplyByConst(deltaTime))
		deltaRangeVector = vector.Create(calculationStep, velocityVector.Y*deltaTime, velocityVector.Z*deltaTime)
		rangeVector = rangeVector.Add(deltaRangeVector)
		velocity = velocityVector.Magnitude()
		time = time + deltaRangeVector.Magnitude()/velocity
	}
	return ranges
}

func calculateStabilityCoefficient(ammunitionInfo Ammunition, rifleInfo Weapon, atmosphere Atmosphere) float64 {
	var weight float64 = ammunitionInfo.Bullet().BulletWeight().In(unit.WeightGrain)
	var diameter float64 = ammunitionInfo.Bullet().BulletDiameter().In(unit.DistanceInch)
	var twist float64 = rifleInfo.Twist().Twist().In(unit.DistanceInch) / diameter
	var length float64 = ammunitionInfo.Bullet().BulletLength().In(unit.DistanceInch) / diameter
	var sd = 30 * weight / (math.Pow(twist, 2) * math.Pow(diameter, 3) * length * (1 + math.Pow(length, 2)))
	var fv = math.Pow(ammunitionInfo.MuzzleVelocity().In(unit.VelocityFPS)/2800, 1.0/3.0)

	var ft float64 = atmosphere.Temperature().In(unit.TemperatureFahrenheit)
	var pt float64 = atmosphere.Pressure().In(unit.PressureInHg)
	var ftp = ((ft + 460) / (59 + 460)) * (29.92 / pt)

	return sd * fv * ftp
}

func windToVector(shot ShotParameters, wind WindInfo) vector.Vector {
	var sightCosine = math.Cos(shot.SightAngle().In(unit.AngularRadian))
	var sightSine = math.Sin(shot.SightAngle().In(unit.AngularRadian))
	var cantCosine = math.Cos(shot.CantAngle().In(unit.AngularRadian))
	var cantSine = math.Sin(shot.CantAngle().In(unit.AngularRadian))
	var rangeVelocity = wind.velocity.In(unit.VelocityFPS) * math.Cos(wind.direction.In(unit.AngularRadian))
	var crossComponent = wind.velocity.In(unit.VelocityFPS) * math.Sin(wind.direction.In(unit.AngularRadian))
	var rangeFactor = -rangeVelocity * sightSine
	return vector.Create(rangeVelocity*sightCosine, rangeFactor*cantCosine+crossComponent*cantSine, crossComponent*cantCosine-rangeFactor*cantSine)
}

func getCorrection(distance, offset float64) float64 {
	return math.Atan(offset / distance)
}

func calculateEnergy(bulletWeight, velocity float64) float64 {
	return bulletWeight * math.Pow(velocity, 2) / 450400
}

func calculateOgv(bulletWeight, velocity float64) float64 {
	return math.Pow(bulletWeight, 2) * math.Pow(velocity, 3) * 1.5e-12
}

File: py_ballisticcalc/trajectory_calculator.py
import math
from .bmath import unit, Vector
from .projectile import Ammunition
from .weapon import Weapon, TwistLeft
from .atmosphere import Atmosphere
from .shot_parameters import ShotParameters
from .wind import WindInfo
from .trajectory_data import TrajectoryData, Timespan

cZeroFindingAccuracy = 0.000005
cMinimumVelocity = 50.0
cMaximumDrop = -15000
cMaxIterations = 10
cGravityConstant = -32.17405


class TrajectoryCalculator(object):
    """
    TrajectoryCalculator table is used to calculate the trajectory
    of a projectile shot with the parameters specified
    """

    def __init__(self):
        self._maximum_calculator_step_size = unit.Distance(1, unit.DistanceFoot).validate()

    @property
    def maximum_calculator_step_size(self) -> unit.Distance:
        """
        :return: the maximum size of one calculation iteration.
        """
        return self._maximum_calculator_step_size

    @maximum_calculator_step_size.setter
    def maximum_calculator_step_size(self, value: unit.Distance):
        """
        Sets the maximum size of one calculation iteration.
        As the generic rule, the maximum step of the calculation must not be greater than
        a half of the step used in the short parameter. The smaller value is, the calculation is more precise but
        takes more time to calculate. From practical standpoint the value in range from 0.5 to 5 feet produces
        good enough accuracy.
        :param value: unit.Distance instance
        :return: None
        """
        self._maximum_calculator_step_size = value

    def get_calculation_step(self, step: float) -> float:
        step = step / 2  # do it twice for increased accuracy of velocity calculation and 10 times per step
        maximum_step: float = self._maximum_calculator_step_size.get_in(unit.DistanceFoot)

        if step > maximum_step:
            step_order = int(math.floor(math.log10(step)))
            maximum_order = int(math.floor(math.log10(maximum_step)))
            step = step / math.pow(10, float(step_order - maximum_order + 1))

        return step

    def sight_angle(self, ammunition: Ammunition, weapon: Weapon, atmosphere: Atmosphere) -> unit.Angular:
        """
        The calculated value is to be used as sightAngle parameter of the ShotParameters structure
        :param ammunition: Ammunition instance
        :param weapon: Weapon instance
        :param atmosphere: Atmosphere instance
        :return: sight angle for a rifle with scope height specified and zeroed using the ammo specified at
        the range specified and under the conditions (atmosphere) specified
        """

        calculation_step = self.get_calculation_step(
            unit.Distance(10, weapon.zero.zero_distance.units).validate().get_in(unit.DistanceFoot))

        mach = atmosphere.mach.get_in(unit.VelocityFPS)
        density_factor = atmosphere.density_factor
        muzzle_velocity = ammunition.muzzle_velocity.get_in(unit.VelocityFPS)
        barrel_azimuth = 0.0
        barrel_elevation = 0

        iterations_count = 0

        zero_finding_error = cZeroFindingAccuracy * 2
        gravity_vector = Vector(0, cGravityConstant, 0)
        while zero_finding_error > cZeroFindingAccuracy and iterations_count < cMaxIterations:
            velocity = muzzle_velocity
            time = 0.0

            # x - distance towards target,
            # y - drop and
            # z - windage

            range_vector = Vector(0.0, -weapon.sight_height.get_in(unit.DistanceFoot), 0)
            velocity_vector = Vector(
                math.cos(barrel_elevation) * math.cos(barrel_azimuth),
                math.sin(barrel_elevation),
                math.cos(barrel_elevation) * math.sin(barrel_azimuth)
            ).multiply_by_const(velocity)

            zero_distance: float = weapon.zero.zero_distance.get_in(unit.DistanceFoot)
            maximum_range = zero_distance + calculation_step

            while range_vector.x <= maximum_range:
                if velocity < cMinimumVelocity or range_vector.y < cMaximumDrop:
                    break

                delta_time = calculation_step / velocity_vector.x
                velocity = velocity_vector.magnitude()
                drag = density_factor * velocity * ammunition.bullet.ballistic_coefficient.drag(velocity / mach)
                velocity_vector = velocity_vector.subtract(
                    (
                        velocity_vector.multiply_by_const(drag).subtract(gravity_vector)
                    ).multiply_by_const(delta_time)
                )
                delta_range_vector = Vector(calculation_step, velocity_vector.y * delta_time,
                                            velocity_vector.z * delta_time)
                range_vector = range_vector.add(delta_range_vector)
                velocity = velocity_vector.magnitude()
                time = time + delta_range_vector.magnitude() / velocity
                if math.fabs(range_vector.x - zero_distance) < 0.5 * calculation_step:
                    zero_finding_error = math.fabs(range_vector.y)
                    barrel_elevation = barrel_elevation - range_vector.y / range_vector.x
                    break

                iterations_count += 1
        return unit.Angular(barrel_elevation, unit.AngularRadian).validate()

    def trajectory(self, ammunition: Ammunition, weapon: Weapon, atmosphere: Atmosphere,
                   shot_info: ShotParameters, wind_info: list[WindInfo]) -> list[TrajectoryData]:
        """
        Calculates the trajectory with the parameters specified
        :param ammunition: Ammunition instance
        :param weapon: Weapon instance
        :param atmosphere: Atmosphere instance
        :param shot_info: ShotParameters instance
        :param wind_info: list[WindInfo]
        :return: trajectory with the parameters specified
        """
        range_to: float = shot_info.maximum_distance.get_in(unit.DistanceFoot)
        step: float = shot_info.step.get_in(unit.DistanceFoot)

        calculation_step = self.get_calculation_step(step)

        bullet_weight = ammunition.bullet.bullet_weight.get_in(unit.WeightGrain)

        stability_coefficient = 1.0

        calculate_drift: bool = False

        if weapon.has_twist and ammunition.bullet.has_dimensions:
            stability_coefficient = self.calculate_stability_coefficient(
                ammunition, weapon, atmosphere
            )
            calculate_drift = True

        ranges_length = int(math.floor(range_to / step)) + 1
        ranges = []

        barrel_azimuth = .0
        barrel_elevation = shot_info.sight_angle.get_in(unit.AngularRadian)
        barrel_elevation = barrel_elevation + shot_info.shot_angle.get_in(unit.AngularRadian)
        alt0: float = atmosphere.altitude.get_in(unit.DistanceFoot)

        # Never used in upstream, uncomment on need
        # density_factor, mach = atmosphere.get_density_factor_and_mach_for_altitude(alt0)

        current_wind = 0
        next_wind_range = 1e7

        if len(wind_info) < 1:
            wind_vector = Vector(0, 0, 0)
        else:
            if len(wind_info) > 1:
                next_wind_range = wind_info[0].until_distance.get_in(unit.DistanceFoot)
            wind_vector = self.wind_to_vector(shot_info, wind_info[0])

        muzzle_velocity = ammunition.muzzle_velocity.get_in(unit.VelocityFPS)
        gravity_vector = Vector(0, cGravityConstant, 0)
        velocity = muzzle_velocity
        time = .0

        # x - distance towards target,
        # y - drop and
        # z - windage

        range_vector = Vector(.0, -weapon.sight_height.get_in(unit.DistanceFoot), 0)
        velocity_vector = Vector(math.cos(barrel_elevation) * math.cos(barrel_azimuth), math.sin(barrel_elevation),
                                 math.cos(barrel_elevation) * math.sin(barrel_azimuth)).multiply_by_const(velocity)

        current_item = 0

        maximum_range = range_to
        next_range_distance = 0

        twist_coefficient = .0

        if calculate_drift:
            if weapon.twist.direction == TwistLeft:
                twist_coefficient = 1
            else:
                twist_coefficient = -1

        while range_vector.x <= maximum_range + calculation_step:
            if velocity < cMinimumVelocity or range_vector.y < cMaximumDrop:
                break

            density_factor, mach = atmosphere.get_density_factor_and_mach_for_altitude(alt0 + range_vector.y)

            if range_vector.x >= next_wind_range:
                current_wind += 1
                wind_vector = self.wind_to_vector(shot_info, wind_info[current_wind])

                if current_wind == len(wind_info) - 1:
                    next_wind_range = 1e7
                else:
                    next_wind_range = wind_info[current_wind].until_distance.get_in(unit.DistanceFoot)

            if range_vector.x >= next_range_distance:
                windage: float = range_vector.z
                if calculate_drift:
                    windage += (1.25 * (stability_coefficient + 1.2) * math.pow(time, 1.83) * twist_coefficient) / 12

                drop_adjustment = self.get_correction(range_vector.x, range_vector.y)
                windage_adjustment = self.get_correction(range_vector.x, windage)

                ranges.append(TrajectoryData(
                    time=Timespan(time),
                    travel_distance=unit.Distance(range_vector.x, unit.DistanceFoot).validate(),
                    drop=unit.Distance(range_vector.y, unit.DistanceFoot).validate(),
                    drop_adjustment=unit.Angular(drop_adjustment, unit.AngularRadian).validate(),
                    windage=unit.Distance(windage, unit.DistanceFoot).validate(),
                    windage_adjustment=unit.Angular(windage_adjustment, unit.AngularRadian),
                    velocity=unit.Velocity(velocity, unit.VelocityFPS),
                    mach=velocity / mach,
                    energy=unit.Energy(self.calculate_energy(bullet_weight, velocity),
                                       unit.EnergyFootPound),
                    optimal_game_weight=unit.Weight(self.calculate_ogv(bullet_weight, velocity),
                                                    unit.WeightPound))
                              )

                next_range_distance += step
                current_item += 1
                if current_item == ranges_length:
                    break

            delta_time = calculation_step / velocity_vector.x
            velocity_adjusted = velocity_vector.subtract(wind_vector)
            velocity = velocity_adjusted.magnitude()

            drag = density_factor * velocity * ammunition.bullet.ballistic_coefficient.drag(velocity / mach)
            velocity_vector = velocity_vector.subtract(
                (velocity_adjusted.multiply_by_const(drag).subtract(gravity_vector)).multiply_by_const(delta_time)
            )
            delta_range_vector = Vector(
                calculation_step, velocity_vector.y * delta_time, velocity_vector.z * delta_time
            )
            range_vector = range_vector.add(delta_range_vector)
            velocity = velocity_vector.magnitude()
            time = time + delta_range_vector.magnitude() / velocity

        return ranges

    @staticmethod
    def calculate_stability_coefficient(ammunition_info: Ammunition,
                                        rifle_info: Weapon, atmosphere: Atmosphere) -> float:
        weight: float = ammunition_info.bullet.bullet_weight.get_in(unit.WeightGrain)
        diameter: float = ammunition_info.bullet.bullet_diameter.get_in(unit.DistanceInch)
        twist: float = rifle_info.twist.twist.get_in(unit.DistanceInch) / diameter
        length: float = ammunition_info.bullet.bullet_length.get_in(unit.DistanceInch) / diameter
        sd = 30 * weight / (math.pow(twist, 2) * math.pow(diameter, 3) * length * (1 + math.pow(length, 2)))
        fv = math.pow(ammunition_info.muzzle_velocity.get_in(unit.VelocityFPS) / 2800, 1.0 / 3.0)
        ft: float = atmosphere.temperature.get_in(unit.TemperatureFahrenheit)
        pt: float = atmosphere.pressure.get_in(unit.PressureInHg)
        ftp = ((ft + 460) / (59 + 460)) * (29.92 / pt)

        return sd * fv * ftp

    @staticmethod
    def wind_to_vector(shot: ShotParameters, wind: WindInfo) -> Vector:
        sight_cosine = math.cos(shot.sight_angle.get_in(unit.AngularRadian))
        sight_sine = math.sin(shot.sight_angle.get_in(unit.AngularRadian))
        cant_cosine = math.cos(shot.cant_angle.get_in(unit.AngularRadian))
        cant_sine = math.sin(shot.cant_angle.get_in(unit.AngularRadian))
        range_velocity = wind.velocity.get_in(unit.VelocityFPS) * math.cos(wind.direction.get_in(unit.AngularRadian))
        cross_component = wind.velocity.get_in(unit.VelocityFPS) * math.sin(wind.direction.get_in(unit.AngularRadian))
        range_factor = -range_velocity * sight_sine
        return Vector(range_velocity * sight_cosine,
                      range_factor * cant_cosine + cross_component * cant_sine,
                      cross_component * cant_cosine - range_factor * cant_sine)

    @staticmethod
    def get_correction(distance: float, offset: float) -> [float, None]:
        if distance != 0:
            return math.atan(offset / distance)
        return

    @staticmethod
    def calculate_energy(bullet_weight: float, velocity: float) -> float:
        return bullet_weight * math.pow(velocity, 2) / 450400

    @staticmethod
    def calculate_ogv(bullet_weight: float, velocity: float) -> float:
        return math.pow(bullet_weight, 2) * math.pow(velocity, 3) * 1.5e-12
